tools/build_third_party_libs.py

```diff
@@ -53,7 +68,7 @@
         cxx_flags = "'-std=c++20 -mmacosx-version-min=10.14'"
     elif sys.platform == "linux":
         platform_dir = "Linux"
-        cxx_flags = "'-std=c++20 -stdlib=libc++'"
+        cxx_flags = f"'-std=c++20 -stdlib=libc++ -nostdinc++ -I{linux_libcpp_include_dir} -L{linux_libcpp_lib_dir} -Qunused-arguments'"
```

Comment:
- I understand why `-nostdinc++` would be necessary here. But is it necessary to specify `-stdlib=libc++` when specifying `-nostdinc++`? It seems weird that we would need to specify both.
- I understand that `cxx_compiler` is invoked by this script, but is `c_compiler` ever invoked? If not, we should get rid of it. I'm thinking it might not be, because we never specify `c_flags`.
- Let's not try to suppress warnings in third-party code.
  - It seems like a lot of work to have this policy for all third-party libs, and if it is not worth doing consistently, it is not worth doing at all.
  - I don't mind seeing the warnings, especially when building third-party libs, which is a one-time operation.
  - In general, we don't understand third-party code well enough to decide if we should really be suppressing warnings or not. I know we do this sometimes in our code, e.g., when including third-party headers. But that is because we would see those warnings when building our code, so there is a risk of too many third-party warnings (which we might not care about) hiding warnings generated by our code (which we should definitely care about). In this case, we are not building any of our code, so there is no such danger.
- We should only add these extra flags if the user didn't specify `args.cxx_compiler`.

Reply:
- Based on my experiments using `-DCMAKE_VERBOSE_MAKEFILE=ON`, I have concluded that `c_compiler` is never invoked by this script. 100% of compiled files are compiled with `cxx_compiler`. We should therefore remove `c_compiler` from this script entirely because it will simplify some of the downstream logic.
- Let's include a `--verbose` option that adds `-DCMAKE_VERBOSE_MAKEFILE=ON` to every `cmake` invocation (excluding the `cmake -build` invocations).
- Let's sort the `cmake` arguments consistently as follows: `DCMAKE_CXX_COMPILER`, `DCMAKE_CXX_FLAGS`, `DCMAKE_PLATFORM_SPECIFIC_STUFF`, `DCMAKE_VERBOSE_MAKEFILE`.

Reply:
1. `-nostdinc++` is required as we don't want to use the standard library headers available in the default paths of the system. So, `-nostdinc++` forces cmake to search for the headers in only directories specified by `-I{}`. However, `-stdlib=libc++` is required as the default standard library on linux is `libstdc++` and clang uses it unless `libc++` is specified.
2. I'll perform a test by removing `c_compiler`, and remove it if not needed.
3. It's weird that adding these extra flags to yaml-cpp build throws `argument unused during compilation` warning for `stdlib=libc++`, `-L{linux_libcpp_lib_dir}`, and `-nostdinc++`, and when these flags are removed (and only `-std=c++20 -I{linux_libcpp_include_dir}` is kept), there are linker errors.

Reply:
Cool.